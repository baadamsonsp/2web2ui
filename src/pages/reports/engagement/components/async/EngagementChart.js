import React from 'react';
import { Bar, BarChart, CartesianGrid, LabelList, ReferenceArea, ResponsiveContainer, XAxis, YAxis } from 'recharts';
import { Panel } from '@sparkpost/matchbox';

import { Empty, PanelLoading } from 'src/components';
import { formatNumber, formatPercent } from 'src/helpers/units';
import { safeRate } from 'src/helpers/math';
import styles from './EngagementChart.module.scss';

// Width must be less than 100% to get ResponsiveContainer to be responsive
const DIMENSIONS = { height: 350, width: '99%' };

// Increased top margin to vertically center the chart
const MARGINS = { bottom: 5, left: 5, right: 5, top: 25 };

// Pad between tallest bar and top of the chart to provide room for the LabelList
const TOP_PADDING = { top: 60 };


// NOTE: ReferenceArea was used because they could easily attach to two bars and provide the
// cooridinates and dimmenions to reliable place this label
function PercentLabel({ percentage, width, x, yAxis }) {
  const label = { height: 24, width: 64 };
  const offset = (width / 2) - (label.width / 2);
  const y = yAxis.y + (yAxis.height / yAxis.tickCount) * Math.round(yAxis.tickCount / 2);

  return (
    <g>
      <svg {...label} x={x + offset} y={y - label.height / 2 + 1}>
        <path
          d="M3 1h52c.7 0 1.3.3 1.6.8l6.4 9c.5.7.5 1.7-.1 2.4l-6.3 8c-.4.5-1 .8-1.6.8H3a2 2 0 0 1-2-2V3c0-1.1.9-2 2-2z"
          fill="white"
          stroke="#e1e1e6"
          strokeMiterlimit="10"
          vectorEffect="non-scaling-stroke"
        />
      </svg>
      <text x={x + offset + 5} y={y + 6} fill="#55555a">{percentage}</text>
    </g>
  );
}

export default function EngagementChart({ accepted = 0, clicks = 0, loading, opens = 0, sent = 0 }) {
  const data = [
    { label: 'Sent', value: sent },
    { label: 'Accepted', value: accepted },
    { label: 'Unique Confirmed Opens', value: opens },
    { label: 'Unique Clicks', value: clicks }
  ];

  if (loading) {
    return <PanelLoading />;
  }

  // sent should always be the largest number, so if it is zero all others should be zero
  if (sent === 0) {
    return <Empty message="No engagement to report" />;
  }

  return (
    <Panel sectioned>
      <div className={styles.EngagementChart}>
        <ResponsiveContainer {...DIMENSIONS}>
          <BarChart barCategoryGap="35%" data={data} margin={MARGINS}>
            <CartesianGrid strokeDasharray="4 1" vertical={false} />
            <XAxis dataKey="label" tickLine={false} />
            <YAxis padding={TOP_PADDING} tickLine={false} tickFormatter={formatNumber} />
            <Bar dataKey="value" fill="#37aadc" isAnimationActive={false}>
              <LabelList fill="#55555a" formatter={formatNumber} position="top" />
            </Bar>
            <ReferenceArea
              x1="Sent" x2="Accepted"
              percentage={formatPercent(safeRate(accepted, sent))}
              shape={PercentLabel}
            />
            <ReferenceArea
              x1="Accepted" x2="Unique Confirmed Opens"
              percentage={formatPercent(safeRate(opens, accepted))}
              shape={PercentLabel}
            />
            <ReferenceArea
              x1="Unique Confirmed Opens" x2="Unique Clicks"
              percentage={formatPercent(safeRate(clicks, accepted))}
              shape={PercentLabel}
            />
          </BarChart>
        </ResponsiveContainer>
      </div>
    </Panel>
  );
}
