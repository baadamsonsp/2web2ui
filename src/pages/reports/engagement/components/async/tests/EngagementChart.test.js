import React from 'react';
import { shallow } from 'enzyme';
import cases from 'jest-in-case';
import EngagementChart from '../EngagementChart';

cases('EngagementChart', (props) => {
  const wrapper = shallow(<EngagementChart {...props} />);
  expect(wrapper).toMatchSnapshot();
}, {
  'renders loading': { loading: true },
  'renders empty panel': { loading: false },
  'renders engagement chart': {
    accepted: 40000,
    clicks: 2525,
    loading: false,
    opens: 5000,
    sent: 50000
  }
});
