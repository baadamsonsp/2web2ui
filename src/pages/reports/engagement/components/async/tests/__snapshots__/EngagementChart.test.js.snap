// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EngagementChart renders empty panel 1`] = `
<Empty
  message="No engagement to report"
/>
`;

exports[`EngagementChart renders engagement chart 1`] = `
<Panel
  sectioned={true}
>
  <div>
    <ResponsiveContainer
      debounce={0}
      height={350}
      width="99%"
    >
      <BarChart
        barCategoryGap="35%"
        barGap={4}
        data={
          Array [
            Object {
              "label": "Sent",
              "value": 50000,
            },
            Object {
              "label": "Accepted",
              "value": 40000,
            },
            Object {
              "label": "Unique Confirmed Opens",
              "value": 5000,
            },
            Object {
              "label": "Unique Clicks",
              "value": 2525,
            },
          ]
        }
        layout="horizontal"
        margin={
          Object {
            "bottom": 5,
            "left": 5,
            "right": 5,
            "top": 25,
          }
        }
        reverseStackOrder={false}
        stackOffset="none"
      >
        <CartesianGrid
          fill="none"
          horizontal={true}
          horizontalPoints={Array []}
          stroke="#ccc"
          strokeDasharray="4 1"
          vertical={false}
          verticalPoints={Array []}
        />
        <XAxis
          allowDataOverflow={false}
          allowDecimals={true}
          dataKey="label"
          domain={
            Array [
              0,
              "auto",
            ]
          }
          height={30}
          hide={false}
          mirror={false}
          orientation="bottom"
          padding={
            Object {
              "left": 0,
              "right": 0,
            }
          }
          reversed={false}
          scale="auto"
          tickCount={5}
          tickLine={false}
          type="category"
          width={0}
          xAxisId={0}
        />
        <YAxis
          allowDataOverflow={false}
          allowDecimals={true}
          domain={
            Array [
              0,
              "auto",
            ]
          }
          height={0}
          hide={false}
          mirror={false}
          orientation="left"
          padding={
            Object {
              "top": 60,
            }
          }
          reversed={false}
          scale="auto"
          tickCount={5}
          tickFormatter={[Function]}
          tickLine={false}
          type="number"
          width={60}
          yAxisId={0}
        />
        <Bar
          animationBegin={0}
          animationDuration={400}
          animationEasing="ease"
          data={Array []}
          dataKey="value"
          fill="#37aadc"
          hide={false}
          isAnimationActive={false}
          layout="vertical"
          legendType="rect"
          minPointSize={0}
          onAnimationEnd={[Function]}
          onAnimationStart={[Function]}
          xAxisId={0}
          yAxisId={0}
        >
          <LabelList
            fill="#55555a"
            formatter={[Function]}
            position="top"
            valueAccessor={[Function]}
          />
        </Bar>
        <ReferenceArea
          alwaysShow={false}
          fill="#ccc"
          fillOpacity={0.5}
          isFront={false}
          percentage="80%"
          r={10}
          shape={[Function]}
          stroke="none"
          strokeWidth={1}
          x1="Sent"
          x2="Accepted"
          xAxisId={0}
          yAxisId={0}
        />
        <ReferenceArea
          alwaysShow={false}
          fill="#ccc"
          fillOpacity={0.5}
          isFront={false}
          percentage="10%"
          r={10}
          shape={[Function]}
          stroke="none"
          strokeWidth={1}
          x1="Accepted"
          x2="Unique Confirmed Opens"
          xAxisId={0}
          yAxisId={0}
        />
        <ReferenceArea
          alwaysShow={false}
          fill="#ccc"
          fillOpacity={0.5}
          isFront={false}
          percentage="5.05%"
          r={10}
          shape={[Function]}
          stroke="none"
          strokeWidth={1}
          x1="Unique Confirmed Opens"
          x2="Unique Clicks"
          xAxisId={0}
          yAxisId={0}
        />
      </BarChart>
    </ResponsiveContainer>
  </div>
</Panel>
`;

exports[`EngagementChart renders loading 1`] = `
<PanelLoading
  minHeight="400px"
/>
`;
